add thread details for declarative config

zeitlinger opened the pull request:
Part of https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/14087
Alternative for https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/14564

## Summary

Moves thread details (`thread.id`, `thread.name`) from the `AddThreadDetailsSpanProcessor` to an `InstrumenterCustomizer`-based approach using `ThreadDetailsAttributesExtractor`. This means thread attributes are added via `AttributesExtractor` at instrumentation build time rather than via `SpanProcessor` at span start time.

The existing `AddThreadDetailsSpanProcessor` is still used for non-declarative-config agent mode (where it remains **on by default** via `otel.javaagent.add-thread-details=true`). The new `InstrumenterCustomizer` path is used for declarative config, where it is **off by default** — users must explicitly enable it:

- **Agent (declarative config)**: `thread_details_enabled: true` under the `java_agent` distribution node (default: false)
- **Spring starter (declarative config)**: `thread_details.enabled: true` under `spring_starter` instrumentation config (default: false)
- **Agent (env var / system properties)**: `otel.javaagent.add-thread-details` (default: true, unchanged)

Making it off by default in 3.0 for non-declarative-config users is tracked in https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/15334.

## Approach

- `ThreadDetailsAttributesExtractor` in `instrumentation-api-incubator` captures `thread.id` and `thread.name` on span start
- `InstrumenterCustomizerProvider` implementations (one for agent, one for spring starter) register the extractor when enabled
- Agent provider reads from `AgentDistributionConfig.isThreadDetailsEnabled()`
- Spring starter provider reads from `ExtendedOpenTelemetry.getInstrumentationConfig("spring_starter")`
- Config lives under distribution nodes (not `instrumentation.common`) per SIG feedback — this is controlled by agent/starter, not individual instrumentations
- The `AddThreadDetailsSpanProcessor` registration is removed from `AgentTracerProviderConfigurer` (the processor class itself is retained for non-DC use)

zeitlinger commented on instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:
```diff
@@ -84,6 +93,16 @@ public final class InstrumenterBuilder<REQUEST, RESPONSE> {
+      // incubator module is not available
+      return false;
+    }
+  }
```
it does for tests in the otel instrumentation API, e.g. https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/a87bb018e24008b084a3c09c8ab023d25b0d98a2/instrumentation/opentelemetry-api/opentelemetry-api-1.50/javaagent/src/incubatorTest/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/v1_50/incubator/logs/LoggerTest.java#L41 - essentially all that are for older versions that do not know ExtendedOpenTelemetry yet

```
java.lang.NoClassDefFoundError: io/opentelemetry/api/incubator/ExtendedOpenTelemetry
	at io.opentelemetry.instrumentation.api.instrumenter.InstrumenterBuilder.addThreadDetailsAttributeExtractor(InstrumenterBuilder.java:471)
	at io.opentelemetry.instrumentation.api.instrumenter.InstrumenterBuilder.buildInstrumenter(InstrumenterBuilder.java:319)
	at io.opentelemetry.instrumentation.api.instrumenter.InstrumenterBuilder.buildInstrumenter(InstrumenterBuilder.java:311)
	at io.opentelemetry.instrumentation.testing.TestInstrumenters.<init>(TestInstrumenters.java:41)
	at io.opentelemetry.instrumentation.testing.InstrumentationTestRunner.<init>(InstrumentationTestRunner.java:68)
	at io.opentelemetry.instrumentation.testing.AgentTestRunner.<init>(AgentTestRunner.java:47)
	at io.opentelemetry.instrumentation.testing.AgentTestRunner.<clinit>(AgentTestRunner.java:40)
	at io.opentelemetry.instrumentation.testing.junit.AgentInstrumentationExtension.<init>(AgentInstrumentationExtension.java:35)
	at io.opentelemetry.instrumentation.testing.junit.AgentInstrumentationExtension.create(AgentInstrumentationExtension.java:39)
	at io.opentelemetry.javaagent.instrumentation.opentelemetryapi.v1_50.logs.LoggerTest.<clinit>(LoggerTest.java:34)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
	at java.base/java.lang.reflect.Field.acquireOverrideFieldAccessor(Field.java:1200)
	at java.base/java.lang.reflect.Field.getOverrideFieldAccessor(Field.java:1169)
	at java.base/java.lang.reflect.Field.get(Field.java:444)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:395)
	at java.base/java.util.stream.DistinctOps$1$2.end(DistinctOps.java:168)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:261)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:510)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.ClassNotFoundException: io.opentelemetry.api.incubator.ExtendedOpenTelemetry
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 27 more
```

laurit replied:
Understood. It would be nice to have a comment that explains why this check is needed. Essentially it is a problem in that test. Or test harness uses instrumenter, but opentelemetry-api tests force using old versions of opentelemetry-api that could be missing features. It is not the first time we have run into issues with these tests. The problem could probably also be solved by delaying initialization of `TestInstrumenters`, that logger test shouldn't need it, though there could be some other test where this approach doesn't work.

trask reviewed (commented):
can you update the PR description with a summary of the relationship between this and the existing thread details span processor? and mention it's off by default when using declarative config (and we can consider making it off by default in 3.0 for others)

trask reviewed (approved):
@jaydeluca can you take a look also?

let's also figure out where to document this, ideally in same place as other declarative config "common" attributes

jaydeluca commented on instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:
```diff
@@ -446,6 +466,28 @@ public void setSpanNameExtractor(
 
+  private static <RESPONSE, REQUEST> void addThreadDetailsAttributeExtractor(
+      InstrumenterBuilder<REQUEST, RESPONSE> builder) {
+    if (isIncubator && builder.openTelemetry instanceof ExtendedOpenTelemetry) {
```
it does feel like this deviates quite a bit to the rest of the instrumentation api, which is otherwise fairly generic. With that being said, I can also see the argument that thread details are a pretty "generic" aspect of java instrumentation.

I tried to experiment with Lauri's suggestion, but ran into some issues with getting it to work with the spring starter, i think the customizers end up running before the spring properties get mapped, so the providers are already invoked.

trask replied:
> Perhaps we should maintain the current behavior and only add it for agent/spring starter?

thinking more about this...

> we also want native instrumentations to be able to use this logic

I'm trying to remember what I was thinking that led us to this...

because native instrumentations would still get thread attributes stamped if the user is using the agent/spring starter (where we register some kind of customizer hook)

zeitlinger replied:
something like this?

```java
        Instrumenter.builder(..)
            .withThreadDetails()
            .addAttributesExtractor(RpcServerAttributesExtractor.create(rpcAttributesGetter))
```

trask replied:
no, this: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/instrumenter/InstrumenterCustomizer.java

zeitlinger replied:
@trask I gave it a shot.

Libraries who want to use thread details now have to 

- add the thread customizer - or add the SDK autoconfigure support that adds the customizer by default
- users still have to opt in using the same flag as agent and starter users

zeitlinger commented:
@trask please have another look

zeitlinger commented on instrumentation/kafka/kafka-connect-2.6/testing/src/test/java/io/opentelemetry/instrumentation/kafkaconnect/v2_6/MongoKafkaConnectSinkTaskTest.java:
```diff
@@ -155,9 +152,7 @@ void testSingleMessage() throws Exception {
                             equalTo(MESSAGING_DESTINATION_NAME, testTopicName),
                             equalTo(MESSAGING_OPERATION, PROCESS),
-                            equalTo(MESSAGING_SYSTEM, KAFKA),
-                            satisfies(THREAD_ID, val -> val.isNotZero()),
```
- thread ID still shows up in the smoke test showing it's not broken: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/24b97e3933fe659e20e7a453734a15a6ca225b87/smoke-tests/src/test/java/io/opentelemetry/smoketest/SpringBootSmokeTest.java#L46
- why not in this test?
  - otel is not `ExtendedOpenTelemetry` - which is done in `OpenTelenetryInstaller` (not part of test support)
  - `ThreadDetailsInstrumenterCustomizerProvider` is not loaded - it's loaded from `sdk-autoconfigure-support` (not part of test support)

zeitlinger replied:
ignore last comment - it's working in this test

jaydeluca commented on instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/thread/ThreadDetailsAttributesExtractor.java:
```diff
@@ -0,0 +1,37 @@
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+import javax.annotation.Nullable;
+
+public class ThreadDetailsAttributesExtractor<RESPONSE, REQUEST>
```
```suggestion
public class ThreadDetailsAttributesExtractor<REQUEST, RESPONSE>
```

trask commented:
discussed in Java declarative config SIG: let's move configuration under java agent and spring starter nodes, since this is something controlled by them as opposed to controlled by individual instrumentations

zeitlinger committed:
move thread details from SpanProcessor to InstrumenterCustomizer

Thread details (thread.id, thread.name) are now added via
InstrumenterCustomizer instead of SpanProcessor, allowing native
instrumentations to opt in.

Configuration moves under distribution nodes:
- javaagent: distribution.javaagent.thread_details_enabled (DC) /
  otel.javaagent.add-thread-details (env var, default true)
- spring starter: instrumentation.spring_starter.thread_details.enabled
  (DC, default false)
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Fix spotless formatting in ThreadDetailsInstrumenterCustomizerProviderTest
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Remove non-Configuration classes from Spring auto-config registries

Java8RuntimeMetricsProvider and Java17RuntimeMetricsProvider are not
@Configuration classes and should not be registered as Spring
auto-configurations. Their presence causes Spring AOT processing to
fail with FileNotFoundException when the classes are not on the
classpath.

Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Fix runtimemetrics -> runtimetelemetry package typo in Spring auto-config registries
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Add java.agent -> otel.javaagent mapping in resolvePropertyKey
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Add bridge mapping for spring_starter.thread_details.enabled

The Spring Boot starter's ThreadDetailsInstrumenterCustomizerProvider reads
from the declarative config path spring_starter > thread_details > enabled,
but the ConfigPropertiesBackedDeclarativeConfigProperties bridge only handles
paths starting with "java." or explicitly listed in SPECIAL_MAPPINGS. This
caused the property otel.instrumentation.common.thread-details.enabled to
have no effect in non-declarative-config mode, breaking smoke tests.
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Read thread details config from Spring Environment instead of broken ExtendedOpenTelemetry path

Replace SPI-based ThreadDetailsInstrumenterCustomizerProvider with
ThreadDetailsAutoConfiguration that reads config directly from Spring
Environment, branching on declarative config vs properties-based like
EarlyConfig. Programmatically registers the customizer provider - no
static mutable fields needed.
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Set thread details flag in OpenTelemetryAutoConfiguration#openTelemetry to fix timing

The previous ThreadDetailsAutoConfiguration ran too late — Spring creates
beans on-demand via DI, so instrumenters were already built before the
flag was set. Moving it into the openTelemetry() bean methods guarantees
correct ordering.
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

trask commented on javaagent-tooling/src/test/java/io/opentelemetry/javaagent/tooling/config/ThreadDetailsInstrumenterCustomizerProviderTest.java:
```diff
@@ -0,0 +1,79 @@
+                span ->
+                    span.hasAttributesSatisfying(
+                        satisfies(THREAD_ID, n -> n.isNotNull()),
+                        satisfies(THREAD_NAME, n -> n.isNotBlank()))),
```
[Testing] `n.isNotNull()` here asserts that the `AbstractLongAssert` object is not null — which is always true — not that `THREAD_ID` has a meaningful value. This passes even if the extractor wrote `0` or `-1`. Prefer a value assertion such as `isPositive()` or `isEqualTo(Thread.currentThread().threadId())`.

```suggestion
                        satisfies(THREAD_ID, n -> n.isEqualTo(Thread.currentThread().getId())),
                        satisfies(THREAD_NAME, n -> n.isEqualTo(Thread.currentThread().getName())))),
```

zeitlinger replied:
Addressed in the current head. The test now asserts exact `thread.id` / `thread.name` values when enabled and `Attributes.empty()` when disabled, so it no longer has the false-positive assertion behavior from the earlier revision.

trask commented on javaagent-tooling/src/test/java/io/opentelemetry/javaagent/tooling/config/ThreadDetailsInstrumenterCustomizerProviderTest.java:
```diff
@@ -0,0 +1,79 @@
+                        satisfies(THREAD_ID, n -> n.isNull()),
+                        satisfies(THREAD_NAME, n -> n.isNull()))));
+  }
+
```
[Testing] When the extractor is disabled, the `thread.id` / `thread.name` attributes should be **absent** from the span, but `satisfies(THREAD_ID, ...)` still expects the key to be present in the attribute set. This doesn't verify what the test name implies. Consider replacing with a positive assertion on the full attribute set, e.g. `span.hasAttributes(Attributes.empty())` (or whatever the expected exact set is), so that an accidentally-emitted `thread.id` actually fails the test.

trask commented on javaagent-tooling/src/test/java/io/opentelemetry/javaagent/tooling/config/ThreadDetailsInstrumenterCustomizerProviderTest.java:
```diff
@@ -0,0 +1,79 @@
+  void enabled(boolean enabled, Consumer<SpanDataAssert> spanAttributesConsumer) {
+    AgentDistributionConfig.resetForTest();
+    AgentDistributionConfig.set(
+        AgentDistributionConfig.fromConfigProperties(
```
[Testing] `resetForTest()` is already invoked from `@AfterEach tearDown()`, so this call is redundant for every iteration after the first. Either drop this line or drop the `@AfterEach` hook — doing both obscures which one is actually establishing preconditions.

```suggestion
  void enabled(boolean enabled, Consumer<SpanDataAssert> spanAttributesConsumer) {
```

trask commented on instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/internal/instrumentation/thread/ThreadDetailsInstrumenterCustomizerProvider.java:
```diff
@@ -0,0 +1,47 @@
+    enabled =
+        Boolean.TRUE.equals(
+            environment.getProperty(
+                "otel.instrumentation.common.thread-details.enabled", Boolean.class));
```
[Config] The two configure methods read from two completely different property names for what appears to be the same feature flag:

- `configureDeclarativeConfig` → `otel.distribution.spring_starter.thread_details_enabled`
- `configureProperties` → `otel.instrumentation.common.thread-details.enabled`

Since both paths read from Spring `Environment` (which already merges system properties, env vars, application.yaml, etc.), is the distinction intentional? The previous `ThreadDetailsAutoConfiguration` gated this with `@ConditionalOnEnabledInstrumentation(module = "common.thread-details")`, i.e. a single `otel.instrumentation.common.thread-details.enabled` toggle. Using a separate `otel.distribution.spring_starter.*` key for declarative config creates an inconsistent user experience compared to the rest of the spring starter and to the javaagent (which uses `otel.javaagent.add-thread-details`). If the two modes really need to diverge, a comment explaining why would help; otherwise please pick one name.

zeitlinger replied:
I kept the existing split in semantics, but made the provider tolerant of both keys so the feature flag resolves coherently either way. Declarative config still prefers the distro-scoped `otel.distribution.spring_starter.thread_details_enabled`, while properties-based config still prefers the legacy `otel.instrumentation.common.thread-details.enabled` toggle for compatibility with the previous spring starter behavior.

trask commented on instrumentation-api-incubator/src/main/java/io/opentelemetry/instrumentation/api/incubator/instrumenter/InstrumenterCustomizer.java:
```diff
@@ -34,6 +35,13 @@ public interface InstrumenterCustomizer {
+   *
+   * @return the OpenTelemetry instance
+   */
+  OpenTelemetry getOpenTelemetry();
```
[API] `getOpenTelemetry()` is added to the incubator `InstrumenterCustomizer` interface (and propagated through `InternalInstrumenterCustomizer`, `InstrumenterCustomizerImpl`, and `InstrumenterBuilder`) but isn't used by any of the new thread-details customizers in this PR. If it isn't needed for this change, consider dropping it and adding it in the PR that actually has a consumer — otherwise the incubator surface grows without a driving use case.

trask commented on declarative-config-bridge/src/main/java/io/opentelemetry/instrumentation/config/bridge/ConfigPropertiesBackedDeclarativeConfigProperties.java:
```diff
@@ -225,6 +225,11 @@ private String resolvePropertyKey(String name) {
       translatedPath.append(translateName(segments[i]));
     }
 
+    // java.agent.* maps to otel.javaagent.* (not otel.instrumentation.agent.*)
+    if (segments.length > 0 && segments[0].equals("agent")) {
+      return "otel.javaagent." + translatedPath.toString().substring("agent.".length());
+    }
```
[General] Minor: if `segments` is exactly `["agent"]` (i.e. someone asks for `java.agent` with no child), `translatedPath` is `"agent"` and `substring("agent.".length())` = `substring(6)` throws `StringIndexOutOfBoundsException`. Low probability given the caller shape, but this is easy to make robust — either guard on `segments.length > 1` or skip the `"agent"` segment during the loop and drop the substring slicing entirely:

```suggestion
    // java.agent.* maps to otel.javaagent.* (not otel.instrumentation.agent.*)
    if (segments.length > 0 && segments[0].equals("agent")) {
      String suffix = translatedPath.length() > "agent.".length()
          ? translatedPath.substring("agent.".length())
          : "";
      return "otel.javaagent." + suffix;
    }
```

trask commented on javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/instrumentation/internal/AgentDistributionConfig.java:
```diff
@@ -226,6 +234,7 @@ private static final class ConfigPropertiesAgentDistributionConfig
               "otel.javaagent.experimental.force-synchronous-agent-listeners", false),
           configProperties.getList("otel.javaagent.exclude-classes"),
           configProperties.getList("otel.javaagent.exclude-class-loaders"),
+          configProperties.getBoolean("otel.javaagent.add-thread-details", true),
```
[Config] The previous default was `!v3Preview` (read from `otel.instrumentation.common.v3-preview`), so users opted into v3 preview got `add-thread-details` defaulted to **off**. The new default is unconditionally `true`, which silently turns `thread.id`/`thread.name` back on for those users. `v3-preview` is still honored elsewhere in the codebase (e.g. `SemconvStability`, `CommonConfig#isV3Preview`). Is this default flip intentional? If so, a CHANGELOG entry calling it out would be useful; if not, please restore the `!v3Preview` default.

trask commented:
hey @zeitlinger, the above review is from AI (based on proposal in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/17889)

zeitlinger committed:
Address remaining thread-details review comments
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

trask committed:
Merge remote-tracking branch 'upstream/main' into declarative-config-thread-details-provider2

# Conflicts:
#	javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/instrumentation/internal/AgentDistributionConfig.java

trask committed:
spotless

trask committed:
Read declarative thread_details flag from ConfigProvider

- configureDeclarativeConfig now takes ConfigProvider and reads
  java.spring_starter.thread_details.enabled from the instrumentation config.
  The call site in OpenTelemetryAutoConfiguration passes SpringConfigProvider,
  which coerces Spring's flattened String values back to Boolean; the previous
  attempts via SdkConfigProvider/YamlDeclarativeConfigProperties silently dropped
  the flag because Spring's application.yaml pipeline stores every scalar as a
  String and YamlDeclarativeConfigProperties refuses to coerce.
- configureProperties keeps reading otel.instrumentation.common.thread-details.enabled
  from Environment for the properties-based path; no legacy fallback in the
  declarative path.
- Add declarative-config smoke test assertion: enable thread_details in the
  testDeclarativeConfig application.yaml and assert thread.id / thread.name
  appear on the server span.
- Simplify segments[0] == "agent" guard in ConfigPropertiesBackedDeclarativeConfigProperties
  (drop dead segments.length check, work off translatedPath).

trask commented:
@zeitlinger I just sent https://github.com/zeitlinger/opentelemetry-java-instrumentation/pull/13

zeitlinger committed:
Keep declarative thread details distro-scoped
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger commented:
I put up an alternative follow-up at https://github.com/zeitlinger/opentelemetry-java-instrumentation/pull/14.

The main reason I didn’t take the config relocation from #13 is that we had already decided in the declarative config SIG to keep this setting under the distro-controlled nodes, not under `instrumentation/development/java/...`. You called that out here: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/15209#issuecomment-2608939609

So #14 keeps the existing spring starter declarative destination `otel.distribution.spring_starter.thread_details_enabled`, but salvages the useful parts of #13:
- adds declarative-config smoke coverage that enables the distro-scoped key from `application.yaml` and asserts `thread.id` / `thread.name`
- takes the unrelated `java.agent` translation robustness fix

I did not take the `ConfigProvider` rewrite because, as implemented today, `getInstrumentationConfig()` only exposes `instrumentation/development`, not the distro node, so that approach only works together with the relocation.

trask committed:
Merge branch 'main' into declarative-config-thread-details-provider2

trask commented:
> keep this setting under the distro-controlled nodes

good point, I'll do a final pass after you merge those updates 👍

zeitlinger committed:
Merge pull request #14 from zeitlinger/pr-15209-alt

Keep declarative thread details distro-scoped

zeitlinger commented:
@copilot resolve the merge conflicts in this pull request

trask committed:
Merge remote-tracking branch 'upstream/main' into declarative-config-thread-details-provider2

# Conflicts:
#	smoke-tests-otel-starter/spring-boot-2/src/testDeclarativeConfig/java/io/opentelemetry/spring/smoketest/OtelSpringStarterSmokeTest.java

trask committed:
Merge branch 'declarative-config-thread-details-provider2' of https://github.com/trask/opentelemetry-java-instrumentation into declarative-config-thread-details-provider2

# Conflicts:
#	declarative-config-bridge/src/main/java/io/opentelemetry/instrumentation/config/bridge/ConfigPropertiesBackedDeclarativeConfigProperties.java
#	smoke-tests-otel-starter/spring-boot-2/src/testDeclarativeConfig/java/io/opentelemetry/spring/smoketest/OtelSpringStarterSmokeTest.java

trask commented:
sorry, more thoughts on this.

I think it's a breaking change and we should hide it behind the v3 preview flag.

and/or maybe we should do both, add the thread attributes in the instrumenter, but also add them as a fallback in the span processor in order to support manual instrumentation done outside of the instrumenter

zeitlinger commented:
> I think it's a breaking change and we should hide it behind the v3 preview flag.

It should not be breaking - but maybe I miss something.

Or do you mean to support https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/16459 - which is releated?

trask commented:
> It should not be breaking - but maybe I miss something.

I was thinking this could break Java agent users who currently get `thread.id` and `thread.name` on manually-created spans?

zeitlinger committed:
Preserve thread-details fallback behavior
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger commented:
Pushed a follow-up in 05c9b20022.

I kept the new instrumenter-based path for declarative config, but restored the span processor as a fallback for spans created outside `Instrumenter`.

To avoid overlap, the processor now backs off when `thread.id` / `thread.name` are already present, so instrumented spans keep the extractor as the primary path while manual spans still get the attributes.

I also fixed the spring declarative-config path to read the distro-scoped `distribution.spring_starter.thread_details_enabled` setting directly, and removed the extra workaround key from the smoke config.

Added test coverage for both processor cases:
- adds thread attrs when missing
- does not overwrite them when already set

zeitlinger committed:
Fix spring thread-details config parsing
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Fix spotless for spring thread-details provider
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Document thread-details preview behavior
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Refactor thread-details preview default
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Inline thread-details preview default
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
Avoid spring config model compile dependency
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

zeitlinger committed:
fix: accept string spring thread-details flag
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

trask committed:
Run spotless

zeitlinger commented:
@trask can this be merged now?

trask commented:
I decided against moving thread attributes from span processor to instrumenter due to losing coverage for manual spans. I think losing coverage for manual spans may be worse (or at least no better) than the current situation of samplers not having access to thread attributes. See new PR #19008.

zeitlinger commented:
Close in favor of https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/19008
